cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -286,6 +316,61 @@ private void deployApp(SystemApplication application) throws Exception {
                  application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
   }
 
+  @VisibleForTesting
+  void autoInstallResources(String capability, List<URL> hubs) {
+    if (hubs == null || hubs.isEmpty()) {
+      LOG.debug("Capability {} do not have auto install resources associated with it", capability);
+      return;
+    }
+
+    for (URL hub : hubs) {
+      HubPackage[] hubPackages;
+      try {
+        URL url = new URL(hub, "/v2/packages.json");
+        String packagesJson = HttpClients.doGetAsString(url);
+        // Deserialize packages.json from hub
+        // See https://cdap.atlassian.net/wiki/spaces/DOCS/pages/554401840/Hub+API?src=search#Get-Hub-Catalog
+        hubPackages = GSON.fromJson(packagesJson, HubPackage[].class);
+      } catch (Exception e) {
+        LOG.warn("Failed to get packages.json from {}. Ignoring error.", hub, e);
+        return;
```

Comment:
Why return? Shouldn't be `continue` to the next hub url?

Reply:
yes, it's better to continue here. fixed.